File: system-test/dotnet/SystemTests/Legacy/Mod03/E2eTests/PlaceOrderNegativeApiTest.cs
using System.Text;
using System.Text.Json;
using SystemTests.Commons.Constants;
using SystemTests.Legacy.Mod03.E2eTests.Base;
using Shouldly;
using Xunit;

namespace SystemTests.Legacy.Mod03.E2eTests;

public class PlaceOrderNegativeApiTest : BaseE2eTest
{
    protected override Task SetShopRawAsync()
    {
        SetUpShopHttpClient();
        return Task.CompletedTask;
    }

    [Fact]
    public async Task ShouldRejectOrderWithNonIntegerQuantity()
    {
        var placeOrderJson = $$"""{"sku":"{{CreateUniqueSku(Defaults.SKU)}}","quantity":"invalid-quantity"}""";

        var uri = new Uri(_configuration.ShopApiBaseUrl + "/api/orders");
        var content = new StringContent(placeOrderJson, Encoding.UTF8, "application/json");
        var response = await _shopApiHttpClient!.PostAsync(uri, content);

        ((int)response.StatusCode).ShouldBe(422);
        var body = await response.Content.ReadAsStringAsync();
        using var doc = JsonDocument.Parse(body);
        var root = doc.RootElement;
        root.GetProperty("detail").GetString().ShouldBe("The request contains one or more validation errors");
        var errors = root.GetProperty("errors");
        errors.GetArrayLength().ShouldBeGreaterThan(0);
        var found = false;
        foreach (var err in errors.EnumerateArray())
        {
            if (err.GetProperty("field").GetString() == "quantity" && err.GetProperty("message").GetString() == "Quantity must be an integer")
            {
                found = true;
                break;
            }
        }
        found.ShouldBeTrue();
    }
}

File: system-test/java/src/test/java/com/optivem/shop/systemtest/legacy/mod03/e2e/PlaceOrderNegativeApiTest.java
package com.optivem.shop.systemtest.legacy.mod03.e2e;

import com.optivem.shop.systemtest.legacy.mod03.e2e.base.BaseE2eTest;
import org.junit.jupiter.api.Test;

import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.UUID;

import static com.optivem.shop.systemtest.commons.constants.Defaults.*;
import static org.assertj.core.api.Assertions.assertThat;

class PlaceOrderNegativeApiTest extends BaseE2eTest {
    @Override
    protected void setShopClient() {
        setUpShopHttpClient();
    }

    @Test
    void shouldRejectOrderWithNonIntegerQuantity() throws Exception {
        var placeOrderJson = """
                {
                    "sku": "%s",
                    "quantity": "invalid-quantity",
                    "country": "%s"
                }
                """.formatted(SKU + "-" + UUID.randomUUID().toString().substring(0, 8), COUNTRY);

        var response = shopApiHttpClient.send(
                HttpRequest.newBuilder()
                        .uri(URI.create(getShopApiBaseUrl() + "/api/orders"))
                        .header("Content-Type", "application/json")
                        .POST(HttpRequest.BodyPublishers.ofString(placeOrderJson))
                        .build(),
                HttpResponse.BodyHandlers.ofString());

        assertThat(response.statusCode()).isEqualTo(422);
        var errorBody = httpObjectMapper.readTree(response.body());
        assertThat(errorBody.get("detail").asText()).isEqualTo("The request contains one or more validation errors");
        var errors = errorBody.get("errors");
        assertThat(errors).isNotNull();
        assertThat(errors.isArray()).isTrue();
        boolean found = false;
        for (var error : errors) {
            if (error.get("field").asText().equals("quantity") && error.get("message").asText().equals("Quantity must be an integer")) {
                found = true;
                break;
            }
        }
        assertThat(found).isTrue();
    }
}
